### DISCUSSION TITLE
[Chapter2.1-2.4] 프로그램이 실행되었지만, 뭐가 뭔지 하나도 모르겠다

### DISCUSSION THREAD

**Discussion by ddongule** [elegant-functional-2023/secrets-of-computer#5] [General]
감상평을 남겨주세요~!
6/22 11시 전까지!

**Comment by ddongule** [comment]
감상평: 
1장에서 느꼈던 감동을 아직 붙잡고 있긴 하다. (^^) 
뭐랄까 진짜 어려운 개념을 쉽게 설명해주려고 무던히 노력하고있는게 느껴진달까? ㅎ.. (이해는 나의 몫이겠지)
아직 읽기만하고 정리는 못했는데, 정리하면 조금 더 와닿을 수 있을 것 같다.
비전공자로서 운영체제, CPU, 쓰레드 등의 개념들에 대해 막연히 무섭다고만 느끼고 있었는데 그 생각을 조금 덜어줬다.
나도 어디가서 누가 물어봤을 때 설명할 수 있을 정도가 되고싶다. (`혹시 이 책을 다 읽을때쯤엔? 가능할까?` 야너두~)
아, 그리고 뭔가 번역가님이 번역도 잘하시는데 약간 애매한 말이다 싶으면 옆에 영어 달아주셔서 넘 좋다.

**Reply by hanana1253** [reply]
ㅋㅋㅋ 완전 공감....

**Reply by sunhpark42** [reply]
강교수님 기대합니다

**Comment by yunseorim1116** [comment]
# 느낀점

2장을 읽으면서 1장과 함께 복습했다. cpu는 하나의 일만 할수 있는데 어떻게 동시에 실행되는것처럼 보이는가? 에 대해 자세히 관심을 가져본적이 없다. 전환 빈도를 빠르게 만들어 동시에 실행되는것처럼 보이게 한다. 그와 동시에 동시성, 병령성에 대해서 정확하게 인지할수 있는 계기가 되었다.

운영체제는 코드의 재사용의 끝판왕. 운영체제가 존재함으로써 pc의 내부 구현 정확하게 알지 못해도 컴퓨터를 편리하게 사용할수 있구나. 
프로세스에 대한 설명을 먼저 하고 그 후에 경량 프로세스, 스레드로 넘어가는 설명이 이해하기 쉬웠다. 프로세스 안에 스레드가 속해 있고 프로세스가 스레드보다 무거운 이유를 합리적이고 납득 가능하게 설명을 해준것같다.

자바스크립트로 개발하는 개발자가 코루틴을 이해하기 위해서는 async await, 프로미스, 제너레이터에 대입해서 공부해도 되는 걸까? 하는 고민이 들었다. 책을 읽은 후에 블로그 글을 좀 더 찾아보았다.

https://charming-kyu.tistory.com/58

[[https://velog.io/@rohkorea86/Generator-함수를-이해해보자-이론편-왜-제네레이터-함수를-써야-하는가](https://velog.io/@rohkorea86/Generator-%ED%95%A8%EC%88%98%EB%A5%BC-%EC%9D%B4%ED%95%B4%ED%95%B4%EB%B3%B4%EC%9E%90-%EC%9D%B4%EB%A1%A0%ED%8E%B8-%EC%99%9C-%EC%A0%9C%EB%84%A4%EB%A0%88%EC%9D%B4%ED%84%B0-%ED%95%A8%EC%88%98%EB%A5%BC-%EC%8D%A8%EC%95%BC-%ED%95%98%EB%8A%94%EA%B0%80)](https://velog.io/@rohkorea86/Generator-%ED%95%A8%EC%88%98%EB%A5%BC-%EC%9D%B4%ED%95%B4%ED%95%B4%EB%B3%B4%EC%9E%90-%EC%9D%B4%EB%A1%A0%ED%8E%B8-%EC%99%9C-%EC%A0%9C%EB%84%A4%EB%A0%88%EC%9D%B4%ED%84%B0-%ED%95%A8%EC%88%98%EB%A5%BC-%EC%8D%A8%EC%95%BC-%ED%95%98%EB%8A%94%EA%B0%80)

**Comment by hanana1253** [comment]
프로세스와 스레드, mutual exclusion 등에 대해서 면접대비하면서 달달 외울 때에는 몰랐던, 실제 동작원리를 알게 되니 신기했다. 특히 왜 스레드에서 스택 영역은 각각 가지고 있는 그림으로 표현했는지 3년만에 이제야 깨우쳤다..ㅎ..

모르면 마법 알면 기술인 컴퓨터의 세계가 나에게 점점 마법의 영역에서 기술의 영역으로 이동하는 듯... 나도 이 책의 친절하면서도 딥다이브한 설명 and 번역에 감동을 받고 있다.

읽기라도 끝까지 하려다 보니 정리는 중간에 포기했지만 복습하면서 꼭 하겠다...
https://teal-line-f3e.notion.site/20240622-2-b12750e6ede7459aa12b353563e53d95

시간에 쫓겨 급하게 하지말고 조금 더 천천히 정독하면 참 좋을 것을^^.. 
금요일 밤에 시작하는 나는 언제쯤 각성할까...

**Reply by ella-yschoi** [reply]
금요일 퇴근 후까지 고생하셨어용
그런데 노션 링크가 접근이 안되어요! 후엥

**Reply by hanana1253** [reply]
엇 이런.. 링크 변경했어요 헤헤..

**Reply by hanana1253** [reply]
p.118 싱글톤 패턴 구현하는 예제의 의도를 모르겠다...

**Comment by ella-yschoi** [comment]
## 감상평

- 늘 느끼지만 이 책은 질문을 던져주고 그에 대한 답변 형식으로 설명해 주어서 더 이해가 잘 되는 것 같아요.
- 이전에 주현 님과 CS 스터디를 했는데 그때는 개념이 낯설어 어렵게 느껴졌어요. 예를 들어 레지스터, 스레드, 블로킹 등.. 그러고나서 이 책을 읽으니 '**아 이게 그거였구나**' 하며 이해가 더 잘 되는 느낌입니다.
- 프론트엔드 개발자로서 자바스크립트 개념과 연결지으며 공부하니 더 재밌었어요. 이때도 역시나 '아 이게 여기서 비롯된건가보다'

<br/>

## 나누고 싶은 이야기

### 2.1장

- 스레드 부분을 읽으면서 마치 주방 안에 **여러 명의 요리사들이 일하는 모습**이 상상됐어요. 프로세스라는 주방 안에, 스레드라는 요리사들이 일하면서, 칼과 재료라는 공유 자원을 가지고 작업을 분담하는거죠. 👩🏻‍🍳
- 스레드 풀 동작 방식은 스레드 여러 개를 미리 생성해두기에 이게 오히려 더 리소스 낭비이지 않을까 생각했어요. 그런데 책에서 말하길 스레드 풀 방식 덕분에 오히려 불필요하게 많은 메모리를 소비하지 않는다고 하더라고요.
- 생각해보면 오히려 필요할 때마다 스레드를 생성하고 소멸시키면 메모리 할당과 해제를 반복하니 **오버헤드가 발생**할 것 같긴 해요. 비유하자면 요리사를 필요할 때마다 고용하면 채용에 너무 많은 리소스를 쓰는 반면, 미리 요리사들을 고용해두고 필요 없을 때는 대기 발령 상태로 두면 더 효율적인 것 같은 느낌..

### 2.2장

- 스레드 전용 저장소를 **TLS**라고 말하더라고요. (어디선가 들어본 것 같기도..)
- 궁금해서 관련 내용을 찾아봤는데 책보다 더 심화 내용(?)을 다룬 [MS의 아티클](https://learn.microsoft.com/en-us/windows/win32/procthread/thread-local-storage)이 있어 공유합니다. 책에 없는 **TLS 인덱스, 슬롯** 이라는 개념도 나와요.

### 2.3장

- 스레드 안전 내용을 읽으며 **순수함수** 개념이 떠올랐어요. 순수함수도 입력 값에 따라 출력값을 계산하는 것처럼, 스레드 안전 또한 스레드가 어떤 순서로 몇 개가 호출되든 서로 간섭하지 않고 올바른 결과를 도출하는 개념이니까요.
- 요약하자면 둘 다 입**출력의 독립성을 가지면서 사이드이펙트가 없다는 게 공통점**이라고 말할 수 있을 듯 합니다.

### 2.4장

- 코루틴의 핵심을 한 줄로 정리하면, 자신의 실행 상태를 저장할 수 있기 때문에 코루틴이 반환된 후에도 계속 호출이 가능하고 마지막에 **일시중지된 지점부터 다시 이어서 실행된다** 군요. 마치 책갈피와 같은 역할이랄까? 🔖
- 책에서 코루틴을 운영 체제가 스레드를 스케줄링하는 것과 비슷하다고 언급했어요. 이에 더해 저는 코루틴을 자바스크립트의 [제너레이터](https://ko.javascript.info/generators), 그리고 [async/await](https://ko.javascript.info/async-await)와 비슷하다고 느꼈어요.
- ES6에서 등장한 제너레이터는 특이하게 `function*` 과 `yield` 키워드를 사용해 함수 실행을 잠시 중지하고 `next()`를 호출하면 중지된 지점에서 다시 실행하죠. 상태를 유지하고 필요할 때마다 실행을 재개할 수 있는 특징이 있어 코루틴과 유사한 것 같아요.
- 잘 아시다시피 ES2017에서 등장한 `async` 함수는 `Promise`를 반환하고, `await`은 `Promise`가 해결될 때까지 함수 실행을 일시 중지하고, 완료되면 다음 줄부터 다시 실행을 이어간다는 점에서 코루틴과 유사한 것 같습니다.
- 위와 같이 Stackoverflow에 저와 같은 궁금증을 가진 분이 있어 [질의응답 내용](https://stackoverflow.com/questions/78171679/relationshipt-between-async-await-generators-and-coroutines)을 공유합니다.
- 더불어 [어떤 블로그 글](https://it-eldorado.tistory.com/159?category=749661)에서 코루틴, 제너레이터, async/await의 관계를 파이썬과 자바스크립트 관점에서 잘 설명해주어서 더 궁금하신 분들은 읽어보셔도 좋을 것 같아요.
- 결론적으로 **제너레이터와 async/await는 코루틴의 개념에서 영감을 받은 기능**이라고 합니다. (두둥)

<br/>

**Comment by thgr8ganzi** [comment]
감상평

1. 원래 알고 있던 내용을 좀 더 깊게 생각해 볼 수 있어 좋았다. 물리적 하드웨어가 아닌 프로세스와 스레드가 논리적인 개념이라는 것을 깨우쳐 줘서 좋았음

2. 최근에 멀티스레딩 환경에서 스레드 풀을 몇 개 정도 설정해야 좋을까라는 생각을 했는데 책에서 N 개의 코의 당 N * (1 + Wait Time / Computing Time)으로 2N으로 설정하라는 레퍼런스를 줘서 좋았음

3. 책에서 코루틴을 자바스크립트의 제너레이터처럼 설명한 것이 있는데 자바스크립트의 yeild 키워드를 사용하는 제너레이터가가 과연 코루틴을 적용한 것일지 아직 의문이 든다.


GPT 답변 : JavaScript의 제너레이터는 코루틴의 일부분을 구현한 것으로 볼 수 있습니다. 제너레이터는 일시 중단과 재개를 지원하며, 이는 코루틴의 중요한 특성 중 하나입니다. 그러나 제너레이터는 코루틴의 완전한 기능을 제공하지 않으며, 주로 단방향 제어 흐름을 지원합니다.
따라서, JavaScript의 제너레이터는 코루틴의 개념을 부분적으로 적용한 것이라고 할 수 있습니다.(할루씨네이션인지 알수 없음)

정리 : https://github.com/thgr8ganzi/TIL/blob/master/CS/secret-of-computer/2.%EC%9A%B4%EC%98%81%EC%B2%B4%EC%A0%9C_%ED%94%84%EB%A1%9C%EC%84%B8%EC%8A%A4_%EC%8A%A4%EB%A0%88%EB%93%9C.md

**Comment by iborymagic** [comment]
감상평

요번 챕터를 읽으면서는 예전에 대학교 때 살짝 발 뒤꿈치정도 담궈봤던 유니티가 떠올랐더랬죠..
스레드 풀에 관한 내용을 읽으면서 '풀(Pool)' 이라는 단어를 어디서 들어봤었나 했더니.. 
유니티의 '오브젝트 풀링(Object Pooling)' 기법에서 들었던 기억이 났더랬죠..
게임 오브젝트 혹은 이펙트같은 것들을 필요할 때마다 생성 및 삭제하면 생성에 사용되는 연산 비용, 
그리고 파괴 이후에 발생하는 가비지 컬렉팅 등으로 인한 프레임 드랍이 발생할 수 있기 때문에 
그것들을 해결하기 위해 미리 오브젝트들을 생성하여 숨겨놓고 재사용하는 기법이 오브젝트 풀링입니다.. 
사실상 스레드 풀과 거의 유사한 느낌이랄까요..

그렇다고 무턱대고 오브젝트 풀링을 막 사용한다기 보다는, 
잔챙이 몬스터들같이 자주 생성되고 자주 사라지는 놈들이나 스킬 이펙트(자주 사용되므로)같은 것들에 사용해야 합니다.. 
재사용될 가능성이 낮은 친구들은 사실상 메모리 낭비에 가깝다고 볼 수 있읍니다..
그래서 오브젝트 풀링은 남용하지 않고 적정한 선에서 잘 사용해줘야 한다는 점.. 
스레드 풀도 적정 개수를 잘 계산하여 사용해줘야 한다는 점..
그렇다는 점..

코루틴도 마찬가지로 유니티에서 처음 배웠던 개념이었더랬죠..
유니티의 코루틴은 저의 짧고 얕은 지식에 의하면.. 애니메이션에 자주 사용되는 개념이었던 것 같읍니다..
보통 함수에서 동작을 수행하면 한 번의 프레임 업데이트 내에 다 실행하려고 하는데.. 
코루틴으로 그걸 각 프레임에 쪼개서 넣을 수 있읍니다..
프레임 1에 찔끔 실행하고 yield, 그리고 프레임 1의 나머지 시간동안은 유니티에게 제어권을 넘겼다가 
프레임 2가 되면 다시 함수로 돌아와서 찔끔 실행하고 yield, 반복반복 요런 식입니다..

많은 분들이 얘기해주셨다시피.. 저도 제너레이터가 바로 떠오를 수밖에 없었는데요..
코루틴은 호출자에서 코루틴으로, 코루틴에서 호출자로 양방향 데이터 전달이 가능한 반면에..
제너레이터는 호출자로부터 데이터를 받는 것만 가능하다는 차이점이 있다고 합니다..
사실 제너레이터를 딱히 쓸 일이 없어서.. 허허.. 별 생각은 안났고.. 
그냥 그렇구나 하고 넘어갔어읍니다..

**Comment by sa02045** [comment]
- 싱글 스레드 환경에서 주로 작업하는 FE개발에게 멀티 스레드 환경이 낯선데 책이 잘 설명해줘서 좋았습니다! 
- 브라우저의 [웹 워커](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API) 같은 개념도 생각이 났어요. [예제](https://github.com/sa02045/js-browser-server-multi-thread-example/blob/main/browser-web-worker/worker.html)도 만들어 봤습니다.
- 코루틴 보면서 다른 분들처럼 저도 generator 생각이 났었습니다. 어떤 차이점이 있는지 궁금했었는데 위에 현철님이 잘 설명해주셨군요!
- 요즘 Next.js 공부하고 있는데 Streaming SSR이 generator와 연관이 있다는 것을 얼핏 봤던 것 같은데 좀 더 공부해봐야겠어요

**Reply by sa02045** [reply]
- 코루틴 (하나의 스레드에서 여러 작업을 수행하기 위해 스위칭하는 개념)
- 제너레이터 (좀 더 문법적인 개념? 여러 개의 값을 필요에 따라 하나씩 반환하는 수단) 

다음주부까지 더 딥하게 알아봐야할 듯

**Reply by sa02045** [reply]
FE 개발자 입장에서 드는 생각
- 무거운 작업과 브라우저 랜더링 블락의 관계
- React 동시성 모드가 등장한 이유?
- Next.js Streaming SSR은 어떻게 구현되는가? Stream을 쪼개서 어떻게 보내는건지?
- 번들링 도구의 차이점에서 스레드, 병렬성 관계
    - 자바스크립트 기반 Wepack은 느리다. go 기반 esbuild는 빠르다. 왜? 멀티코어
 - Node.js Express와 Java Spring의 성능 차이 (싱글스레드 작업, 멀티스레드 작어 )

**Comment by sunhpark42** [comment]
1. 코루틴, 지나가면서 들어보긴했는데 이런거였구나 싶다. Promise/async await도 이렇게 구현되어있으려나 싶었음.
2. 동시성 개념, 예전엔 이해 안됐는데 한 10번쯤 보니 익숙해 진듯.

**Comment by bigsaigon333** [comment]
- 멀티 태스킹에 대한 나름의 멘탈모델은 잡혀있다고 생각했는데, cpu 레벨의 pc 레지스터를 이용하는 실행의 단순함, 컨텍스트를 기반으로 한 메모리 정보의 전환 레이어, 컴파일러/링커 레이어 등 각 부분이 추상화가 잘 되어 있어서, 동시성을 아무 의식하지 않고 컴퓨터를 사용할 수 있는 것을 깨닫게 되었습니다

**Comment by chloe-codes1** [comment]
Thread safe 하면서 성능도 챙기는 프로그램을 만들기 위한 고민은 계속되어 왔지만, 구현체 레벨에서 생각해 왔었는데 좀 더 로우 레벨을 훑어 볼 수 있어 좋았다.

코루틴을 늘 공부해야지 생각만 하고 있었는데 전체적인 개념을 익힐 수 있어 좋았다.
Kotlin과 Go에서 코루틴이 어떤 방식으로 적용되고 활용되었는지 더 공부해야겠다.

**Comment by sunhpark42** [comment]
- 스레드 로컬 저장소에는 사용여부와 상관없이 인스턴스는 만들어진다

by GPT
`__thread` 또는 `thread_local`로 선언한 변수는 각 스레드마다 고유한 인스턴스를 가지게 됩니다. 즉, 각 스레드는 자신의 스레드 로컬 저장소에 이 변수의 인스턴스를 따로 가지게 됩니다. 하지만, 그 변수를 사용하지 않을 때에도 여전히 각 스레드의 스레드 로컬 저장소에 이 변수는 존재합니다.

이러한 스레드 로컬 저장소는 스레드가 생성될 때 할당되며, 스레드가 종료될 때까지 유지됩니다. 따라서, 스레드가 해당 변수를 사용하지 않더라도 메모리 공간이 할당되고 존재하게 됩니다.

여기에 대한 근거자료는 다음과 같습니다:

1. **C++11 표준 (ISO/IEC 14882:2011)**
   - C++11 표준에서는 `thread_local` 키워드를 사용하여 스레드 로컬 변수를 선언할 수 있으며, 이는 스레드별로 고유한 인스턴스를 생성함을 명시하고 있습니다. 표준에서는 이러한 변수들이 스레드의 라이프타임 동안 존재한다고 명시하고 있습니다.
   
2. **POSIX 표준 (IEEE Std 1003.1-2017, POSIX.1-2017)**
   - POSIX 스레드 라이브러리에서는 스레드 로컬 저장소 (Thread-Specific Data, TSD)를 관리하기 위한 API를 제공하며, 이는 각 스레드가 자신의 TSD에 접근할 수 있도록 보장합니다. POSIX 표준에서는 TSD가 스레드가 생성될 때 할당되며, 스레드가 종료될 때 해제된다고 명시하고 있습니다.

3. **GCC 컴파일러 문서**
   - GCC 컴파일러는 `__thread` 키워드를 사용하여 스레드 로컬 변수를 선언할 수 있게 합니다. GCC 문서에서는 `__thread` 변수가 각 스레드마다 고유한 인스턴스를 가지며, 스레드가 종료될 때까지 유지된다고 설명하고 있습니다.

4. **LLVM 및 Clang 문서**
   - LLVM 및 Clang 컴파일러도 `__thread` 및 `thread_local` 키워드를 지원하며, 문서에서는 이들 변수가 각 스레드마다 고유한 인스턴스를 가지며, 스레드가 종료될 때 해제된다고 설명하고 있습니다.

다음은 이러한 내용을 참고할 수 있는 몇 가지 자료들입니다:

- [C++11 표준 문서](https://isocpp.org/std/the-standard)
- [POSIX 표준 문서](https://pubs.opengroup.org/onlinepubs/9699919799/)
- [GCC 문서](https://gcc.gnu.org/onlinedocs/gcc-12.1.0/gcc/Thread-Local.html)
- [LLVM 및 Clang 문서](https://clang.llvm.org/docs/LanguageExtensions.html#thread-local)

이러한 자료들을 통해, `__thread` 또는 `thread_local` 변수는 각 스레드가 고유한 인스턴스를 가지며, 해당 변수를 사용하지 않더라도 각 스레드의 로컬 저장소에 존재함을 알 수 있습니다.

- tls(스레드전용저장소) 에 복사된 데이터 정보가 각 스레드에서 유지돼? 혹은 스레드가 작업을 수행하고 나면 초기화 돼?
ChatGPT
TLS(스레드 전용 저장소)에 복사된 데이터는 각 스레드에서 고유하게 유지돼. 각 스레드는 자신의 TLS를 가지고 있고, 다른 스레드와는 독립적으로 운영돼. 따라서 스레드가 작업을 수행하는 동안 TLS에 저장된 데이터는 유지되고, 스레드가 종료되기 전까지는 초기화되지 않아.

TLS의 주요 장점 중 하나는 여러 스레드가 동시에 실행되더라도 각 스레드가 자신의 데이터를 안전하게 보관할 수 있다는 거야. 즉, 스레드가 종료되거나 명시적으로 초기화하기 전까지는 TLS에 저장된 데이터가 유지돼.

스레드가 종료되는 시점은 언제야?
ChatGPT
스레드가 종료되는 시점은 다음과 같은 경우에 발생해:

스레드 함수가 종료될 때: 스레드가 수행할 작업을 정의한 함수가 반환될 때 스레드는 종료돼. 예를 들어, pthread_create로 생성된 스레드가 start_routine 함수를 실행하고 이 함수가 반환되면 스레드는 종료돼.

스레드가 명시적으로 종료될 때: 스레드가 pthread_exit 같은 함수를 호출하여 명시적으로 자신을 종료할 때. 이 경우, 스레드는 즉시 종료돼.

프로세스가 종료될 때: 프로세스가 종료되면 해당 프로세스에 속한 모든 스레드도 함께 종료돼. 이는 exit 함수가 호출되거나 프로세스가 어떤 이유로든 종료될 때 발생해.

다른 스레드에 의해 종료될 때: 다른 스레드가 pthread_cancel 같은 함수를 호출하여 특정 스레드를 종료시킬 수 있어. 이 경우, 대상 스레드는 종료돼.

이와 같은 시점에 스레드는 종료되며, 이때 TLS에 저장된 데이터는 더 이상 유효하지 않게 돼. 스레드가 종료되면 TLS의 자원도 해제돼.

**Comment by sunhpark42** [comment]
@sa02045 님의 2장 퀴이즈
https://forms.gle/c9gU87Nag1VAyzKm6

**Comment by sunhpark42** [comment]
식사하는 철학자
https://namu.wiki/w/%EC%8B%9D%EC%82%AC%ED%95%98%EB%8A%94%20%EC%B2%A0%ED%95%99%EC%9E%90%20%EB%AC%B8%EC%A0%9C

**Comment by sunhpark42** [comment]
Mutex

- 쉬운 설명
뮤텍스(Mutex)는 "Mutual Exclusion"의 줄임말로, 컴퓨터 프로그래밍에서 여러 스레드나 프로세스가 공유 자원에 접근할 때 발생할 수 있는 충돌을 방지하기 위해 사용되는 도구야. 쉽게 말해, 뮤텍스는 특정 코드나 데이터에 동시에 접근하는 것을 막기 위한 자물쇠와 같은 역할을 해.

뮤텍스의 기본 개념
공유 자원 보호: 여러 스레드가 동시에 공유 자원(예: 데이터베이스, 파일, 변수 등)에 접근하면 데이터가 엉킬 수 있어. 이를 방지하기 위해 뮤텍스를 사용해 한 번에 하나의 스레드만 자원에 접근하도록 해.

잠금(Lock)과 해제(Unlock): 뮤텍스는 두 가지 주요 동작을 제공해:

잠금(Lock): 스레드가 자원에 접근하기 전에 뮤텍스를 잠가 다른 스레드가 접근하지 못하게 해. 자물쇠를 잠그는 것과 비슷해.
해제(Unlock): 스레드가 자원에 대한 작업을 마친 후 뮤텍스를 해제해 다른 스레드가 자원에 접근할 수 있게 해. 자물쇠를 푸는 것과 같아.
예시
예를 들어, 두 명의 사람이 동시에 은행 계좌에서 돈을 출금하려고 한다고 생각해보자. 이때 뮤텍스를 사용하지 않으면 두 명이 동시에 출금을 시도해 계좌 잔액이 잘못 계산될 수 있어. 그러나 뮤텍스를 사용하면 다음과 같이 동작해:

첫 번째 사람이 출금을 시도할 때 뮤텍스를 잠금.
첫 번째 사람이 출금을 완료하고 뮤텍스를 해제.
두 번째 사람이 출금을 시도하고, 첫 번째 사람이 해제한 뮤텍스를 잠금.
두 번째 사람이 출금을 완료하고 뮤텍스를 해제.
이렇게 하면 두 명의 출금 작업이 안전하게 처리돼.

요약
뮤텍스는 자물쇠: 특정 자원이나 코드에 동시에 여러 스레드가 접근하지 못하도록 막아줌.
잠금과 해제: 스레드가 자원에 접근할 때 잠그고, 작업이 끝나면 해제함.
충돌 방지: 여러 스레드가 동시에 자원에 접근할 때 발생할 수 있는 데이터 충돌을 방지해.

- 조금 더 전문적인 설명
뮤텍스(Mutex)는 컴퓨터 과학에서 스레드 간의 동기화를 위한 중요한 도구 중 하나야. 전공생에게 설명할 때는 좀 더 기술적이고 구체적으로 접근할 수 있어.

뮤텍스의 정의 및 목적
뮤텍스는 Mutual Exclusion의 약자로, 다중 스레드 환경에서 공유 자원에 대한 접근을 직렬화하여 데이터 일관성을 보장하기 위한 동기화 원시기야. 뮤텍스를 사용하면 한 번에 하나의 스레드만 특정 코드 섹션(임계 구역, critical section)에 접근할 수 있어.

주요 기능
Lock(잠금): 스레드가 임계 구역에 진입하기 전에 뮤텍스를 잠급니다. 만약 뮤텍스가 이미 잠겨 있으면, 해당 스레드는 다른 스레드가 뮤텍스를 해제할 때까지 대기 상태에 들어갑니다.
Unlock(해제): 스레드가 임계 구역에서 작업을 마친 후 뮤텍스를 해제합니다. 해제된 뮤텍스를 기다리고 있던 다른 스레드 중 하나가 뮤텍스를 획득하고 임계 구역에 진입할 수 있습니다.
동작 원리
뮤텍스는 주로 다음과 같은 방식으로 사용돼:

초기화: 뮤텍스를 사용하기 전에 초기화해야 해. POSIX 스레드의 경우 pthread_mutex_init 함수를 사용해 초기화할 수 있어.
잠금: 스레드가 임계 구역에 들어가기 전 pthread_mutex_lock 함수를 호출해 뮤텍스를 잠가.
작업 수행: 임계 구역 내에서 필요한 작업을 수행해.
해제: 작업이 끝나면 pthread_mutex_unlock 함수를 호출해 뮤텍스를 해제해.
예제 코드 (POSIX 스레드 사용)
```c
#include <pthread.h>
#include <stdio.h>

pthread_mutex_t mutex;
int shared_data = 0;

void* thread_function(void* arg) {
    pthread_mutex_lock(&mutex); // 뮤텍스 잠금
    shared_data++; // 임계 구역
    printf("Shared Data: %d\n", shared_data);
    pthread_mutex_unlock(&mutex); // 뮤텍스 해제
    return NULL;
}

int main() {
    pthread_t threads[10];

    pthread_mutex_init(&mutex, NULL); // 뮤텍스 초기화

    for (int i = 0; i < 10; i++) {
        pthread_create(&threads[i], NULL, thread_function, NULL);
    }

    for (int i = 0; i < 10; i++) {
        pthread_join(threads[i], NULL);
    }

    pthread_mutex_destroy(&mutex); // 뮤텍스 소멸

    return 0;
}
```
이 예제에서는 10개의 스레드가 shared_data 변수를 증가시키는 작업을 수행해. pthread_mutex_lock과 pthread_mutex_unlock을 사용해 뮤텍스를 잠그고 해제함으로써, 동시에 여러 스레드가 shared_data에 접근하는 것을 방지해.

중요한 고려 사항
데드락(Deadlock): 잘못된 뮤텍스 사용으로 인해 데드락이 발생할 수 있어. 이를 방지하기 위해 뮤텍스 잠금 순서를 일관되게 유지하고, 필요한 최소 시간 동안만 잠금을 유지해야 해.
우선순위 역전(Priority Inversion): 높은 우선순위 스레드가 낮은 우선순위 스레드에 의해 블록될 수 있어. 이를 해결하기 위해 우선순위 상속(priority inheritance) 같은 기법을 사용할 수 있어.
이 설명이 좀 더 전공생에게 적합한 내용일 거야. 이해하는 데 도움이 되었길 바라.

**Comment by sunhpark42** [comment]
That is, since generators are primarily used to simplify the writing of [iterators](https://en.wikipedia.org/wiki/Iterator), the yield statement in a generator does not specify a coroutine to jump to, but rather passes a value back to a parent routine.
